### ISSUE TITLE
Using scrcpy with Show Touches switch (-t, --show-touches) throws exceptions

### ISSUE THREAD

**Issue by spncrhrstn** [Genymobile/scrcpy#2788]
- [x] I have read the [FAQ](https://github.com/Genymobile/scrcpy/blob/master/FAQ.md).
 - [x] I have searched in existing [issues](https://github.com/Genymobile/scrcpy/issues).

**Environment**
 - OS: Windows 10
 - scrcpy version: 1.20
 - installation method: Windows release
 - device model: Pixel 4a
 - Android version: 12

**Describe the bug**
On my recently updated to Android 12 Pixel 4a, the switch to show touches (-t, --show-touches) now throws an error. Also tried manually changing the setting in phone developer options and then running scrcpy with the switch. Still throws the same error (though touches are shown).

This issue was present in the last version of scrcpy as well as the current version (1.20). Also worked fine IIRC on Android 11.

Looks like it might be a similar issue to #2671 

Console output:
```
D:\Programs\scrcpy-win64>scrcpy -t
INFO: scrcpy 1.20 <https://github.com/Genymobile/scrcpy>
D:\Programs\scrcpy-win64\scrcpy-server: 1 file pushed, 0 skipped. 69.5 MB/s (37139 bytes in 0.001s)
[server] INFO: Device: Google Pixel 4a (Android 12)
[server] ERROR: Could not invoke method
java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Method.invoke(Native Method)
        at com.genymobile.scrcpy.wrappers.ContentProvider.call(ContentProvider.java:108)
        at com.genymobile.scrcpy.wrappers.ContentProvider.getValue(ContentProvider.java:149)
        at com.genymobile.scrcpy.wrappers.ContentProvider.getAndPutValue(ContentProvider.java:165)
        at com.genymobile.scrcpy.Server.scrcpy(Server.java:32)
        at com.genymobile.scrcpy.Server.main(Server.java:255)
        at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
        at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:355)
Caused by: java.lang.SecurityException: Calling uid: 2000 doesn't match source uid: 0
        at android.os.Parcel.createExceptionOrNull(Parcel.java:2425)
        at android.os.Parcel.createException(Parcel.java:2409)
        at android.os.Parcel.readException(Parcel.java:2392)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:190)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:142)
        at android.content.ContentProviderProxy.call(ContentProviderNative.java:732)
        ... 8 more
Caused by: android.os.RemoteException: Remote stack trace:
        at android.content.AttributionSource.enforceCallingUid(AttributionSource.java:239)
        at android.content.AttributionSource.<init>(AttributionSource.java:158)
        at android.content.AttributionSource$1.createFromParcel(AttributionSource.java:454)
        at android.content.AttributionSource$1.createFromParcel(AttributionSource.java:446)
        at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:289)

[server] ERROR: Could not invoke method
java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Method.invoke(Native Method)
        at com.genymobile.scrcpy.wrappers.ContentProvider.call(ContentProvider.java:108)
        at com.genymobile.scrcpy.wrappers.ContentProvider.putValue(ContentProvider.java:161)
        at com.genymobile.scrcpy.wrappers.ContentProvider.getAndPutValue(ContentProvider.java:167)
        at com.genymobile.scrcpy.Server.scrcpy(Server.java:32)
        at com.genymobile.scrcpy.Server.main(Server.java:255)
        at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
        at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:355)
Caused by: java.lang.SecurityException: Calling uid: 2000 doesn't match source uid: 0
        at android.os.Parcel.createExceptionOrNull(Parcel.java:2425)
        at android.os.Parcel.createException(Parcel.java:2409)
        at android.os.Parcel.readException(Parcel.java:2392)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:190)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:142)
        at android.content.ContentProviderProxy.call(ContentProviderNative.java:732)
        ... 8 more
INFO: Renderer: direct3d
INFO: Initial texture: 1080x2336
```

**Comment by rom1v** [comment]
> Caused by: java.lang.SecurityException: Calling uid: 2000 doesn't match source uid: 0

Is your device rooted? Try `adb unroot` to test.

**Comment by spncrhrstn** [comment]
My device is not rooted. `adb unroot` shows  

```
D:\Programs\scrcpy-win64>adb unroot
adbd not running as root
```

**Comment by rom1v** [comment]
If you enable it manually like this:

```
adb shell settings put system show_touches 1
```

Does it work?

**Comment by spncrhrstn** [comment]
Yep, running that command does enable the show touches setting in developer options. And of course shows touches

**Comment by rom1v** [comment]
OK, it seems there is a change in Android 12 (like #2671).

I will try on a Pixel 4a with Android 12 as soon as I can.

**Comment by rom1v** [comment]
OK, I can reproduce on a Pixel 4a :tada:

I will need to use another way to set system settings, either by calling private APIs to execute system commands in the server process (like [Am.java](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/master/cmds/am/src/com/android/commands/am/Am.java)), either by just calling the executable `am` with arguments from the server process.

I already did the latter some time ago (before implementing the `ContentProvider` wrapper), but on old device executing `am` runs a new Java process, so it was to slow. On Android 12 it will be ok.

It will require some work, but this is now my priority task for the next release :)

**Comment by rom1v** [comment]
Please test #2802.

**Comment by spncrhrstn** [comment]
> Please test #2802.

Tested out the newest files, it works! No errors in the console, and touches are shown on screen.

```
D:\Programs\scrcpy-win64>scrcpy -t
INFO: scrcpy 1.20 <https://github.com/Genymobile/scrcpy>
D:\Programs\scrcpy-win64\scrcpy-server: 1 file pushed, 0 skipped. 44.7 MB/s (38647 bytes in 0.001s)
[server] INFO: Device: Google Pixel 4a (Android 12)
INFO: Renderer: direct3d
INFO: Initial texture: 1080x2336
```

Thanks for taking care of this issue!

**Comment by rom1v** [comment]
:+1: Fixed by #2802.
